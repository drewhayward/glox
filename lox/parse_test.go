package lox

import (
	"fmt"
	"strings"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
)

func TestParseSnapshot(t *testing.T) {
	testCases := []struct {
		source string
	}{
		{`
1;
        `},
		{`
1 + 1;
        `},
		{`
var a = 1;
        `},
		{`
fun foo() {}
        `},
		{`
fun foo() {
print "hello";
}
        `},
		{`
var a = 1;
{
var a = 2;
print a;
}
print a;
        `},
		{`
class Foo {}
        `},
		{`
class Foo {
bar() {}
}
        `},
	}

	for _, tc := range testCases {
		s := snaps.WithConfig()

		t.Run(fmt.Sprintf("Parse(%q)", strings.Trim(tc.source, " \n")), func(t *testing.T) {
			tokens, err := ScanTokens(tc.source)
			if err != nil {
				t.Fatalf(err.Error())
				t.FailNow()
			}

			node, err := Parse(tokens)
			if err != nil {
				t.Fatalf(err.Error())
				t.FailNow()
			}

			s.MatchSnapshot(t, node)
		})
	}
}
