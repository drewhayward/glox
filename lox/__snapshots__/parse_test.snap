
[TestParseSnapshot/Parse("1;") - 1]
lox.ProgramNode{
    Statements: {
        lox.ExprStmt{
            Expr: lox.LiteralExpr[float64]{value:1},
        },
    },
}
---

[TestParseSnapshot/Parse("1_+_1;") - 1]
lox.ProgramNode{
    Statements: {
        lox.ExprStmt{
            Expr: lox.BinaryExpr{
                Operation: "PLUS",
                Lhs:       lox.LiteralExpr[float64]{value:1},
                Rhs:       lox.LiteralExpr[float64]{value:1},
            },
        },
    },
}
---

[TestParseSnapshot/Parse("var_a_=_1;") - 1]
lox.ProgramNode{
    Statements: {
        lox.DeclarationStmt{
            Name: "a",
            Expr: &lox.LiteralExpr[float64]{value:1},
        },
    },
}
---

[TestParseSnapshot/Parse("var_a_=_1;\n{\nvar_a_=_2;\nprint_a;\n}\nprint_a;") - 1]
lox.ProgramNode{
    Statements: {
        lox.DeclarationStmt{
            Name: "a",
            Expr: &lox.LiteralExpr[float64]{value:1},
        },
        lox.BlockStmt{
            Statements: {
                lox.DeclarationStmt{
                    Name: "a",
                    Expr: &lox.LiteralExpr[float64]{value:2},
                },
                lox.PrintStmt{
                    Expr: lox.VarExpr{Name:"a"},
                },
            },
        },
        lox.PrintStmt{
            Expr: lox.VarExpr{Name:"a"},
        },
    },
}
---

[TestParseSnapshot/Parse("fun_foo()_{}") - 1]
lox.ProgramNode{
    Statements: {
        lox.FunctionDeclarationStmt{
            Name:       "foo",
            Parameters: {},
            Body:       lox.BlockStmt{
                Statements: {
                },
            },
        },
    },
}
---

[TestParseSnapshot/Parse("fun_foo()_{\nprint_\"hello\";\n}") - 1]
lox.ProgramNode{
    Statements: {
        lox.FunctionDeclarationStmt{
            Name:       "foo",
            Parameters: {},
            Body:       lox.BlockStmt{
                Statements: {
                    lox.PrintStmt{
                        Expr: lox.LiteralExpr[string]{value:"hello"},
                    },
                },
            },
        },
    },
}
---

[TestParseSnapshot/Parse("class_Foo_{}") - 1]
lox.ProgramNode{
    Statements: {
        lox.ClassDeclarationStmt{
            Name:      "Foo",
            Functions: {
            },
        },
    },
}
---

[TestParseSnapshot/Parse("class_Foo_{\nbar()_{}\n}") - 1]
lox.ProgramNode{
    Statements: {
        lox.ClassDeclarationStmt{
            Name:      "Foo",
            Functions: {
                {
                    Name:       "bar",
                    Parameters: {},
                    Body:       lox.BlockStmt{
                        Statements: {
                        },
                    },
                },
            },
        },
    },
}
---
